cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputFormat.java

```diff
@@ -128,6 +128,9 @@ private static TaskAttemptContext getTaskAttemptContext(TaskAttemptContext conte
                                                           String newOutputName) throws IOException {
     Job job = new Job(context.getConfiguration());
     FileOutputFormat.setOutputName(job, newOutputName);
+    // CDAP-4806 We must set this parameter in addition to calling FileOutputFormat#setOutputName, because
+    // AvroKeyOutputFormat/AvroKeyValueOutputFormat use a different parameter for the output name than FileOutputFormat.
+    job.getConfiguration().set("avro.mo.config.namedOutput", newOutputName);
```

Comment:
Updated to check against the class name string of those two classes to avoid having additional dependencies in this module.